File: ios/DataModels/PrimerFlowRN.swift
//
//  PrimerFlowRN.swift
//  ReactNative
//
//  Created by Carl Eriksson on 10/08/2021.
//  Copyright © 2021 Facebook. All rights reserved.
//
import PrimerSDK

struct PrimerFlowRN: Decodable {
    let intent: PrimerIntentRN
    let paymentMethod: PrimerPaymentMethodTypeRN
    
    
    func toPrimerSessionFlow() -> PrimerSessionFlow? {
        
        switch (intent) {
        case .Checkout:
            switch (paymentMethod) {
            case .Any:
                return .default
            case .ApplePay:
                return .checkoutWithApplePay
            case .Klarna:
                return .checkoutWithKlarna
            case .Card:
                return .completeDirectCheckout
            default:
                return nil
            }
        case .Vault:
            switch (paymentMethod) {
            case .Any:
                return .defaultWithVault
            case .Klarna:
                return .addKlarnaToVault
            case .PayPal:
                return .addPayPalToVault
            case .Card:
                return .defaultWithVault
            default:
                return nil
            }
        }
    }
}

enum PrimerIntentRN: String, Decodable {
    case Checkout, Vault
}

enum PrimerPaymentMethodTypeRN: String, Decodable {
    case `Any`, Klarna, Card, PayPal, GooglePay, ApplePay, GoCardless
}

File: android/src/main/java/com/primerioreactnative/DataModels/PrimerFlowRN.kt
package com.primerioreactnative.DataModels

import io.primer.android.PaymentMethod
import io.primer.android.payment.card.Card
import io.primer.android.payment.gocardless.GoCardless
import io.primer.android.payment.google.GooglePay
import io.primer.android.payment.klarna.Klarna
import io.primer.android.payment.paypal.PayPal
import kotlinx.serialization.Serializable

@Serializable
data class PrimerFlowRN(
  val intent: PrimerIntentRN = PrimerIntentRN.Checkout,
  val paymentMethod: PrimerPaymentMethodTypeRN = PrimerPaymentMethodTypeRN.Any,
) {

  fun toPaymentMethods(): List<PaymentMethod> {
    return when (paymentMethod) {
      PrimerPaymentMethodTypeRN.Any -> listOf(Klarna(), Card(), PayPal())
      PrimerPaymentMethodTypeRN.Klarna -> listOf(Klarna(), Card(), PayPal())
      PrimerPaymentMethodTypeRN.Card -> listOf(Klarna(), Card(), PayPal())
      PrimerPaymentMethodTypeRN.PayPal -> listOf(Klarna(), Card(), PayPal())
      else -> throw Exception()
    }
  }
}

@Serializable
enum class PrimerIntentRN {
  Checkout,
  Vault,
}

@Serializable
enum class PrimerPaymentMethodTypeRN {
  Any,
  Klarna,
  Card,
  PayPal,
  GooglePay,
  ApplePay,
  GoCardless,
}
